hazelcast/src/hazelcast/util/ILogger.cpp

```diff
@@ -115,7 +130,47 @@ namespace hazelcast {
         }
 
         bool ILogger::isFinestEnabled() const {
-            return isEnabled(client::FINEST);
+            return isEnabled(client::LoggerLevel::FINEST);
+        }
+
+        const char *ILogger::getLevelString(client::LoggerLevel::Level logLevel) const {
+            return client::LoggerLevel::getLevelString(logLevel);
+        }
+
+        void ILogger::printMessagePrefix(client::LoggerLevel::Level logLevel) const {
+            char buffer[TIME_STRING_LENGTH];
+            std::cout << getTime(buffer, TIME_STRING_LENGTH) << " " << getLevelString(logLevel) << ": [" << getThreadId()
+                      << "] " << prefix << " ";
+        }
+
+        void ILogger::printLog(client::LoggerLevel::Level level, const std::string &message) {
+            if (!isEnabled(level)) {
+                return;
+            }
+
+            {
+                util::LockGuard l(lockMutex);
+                printMessagePrefix(level);
+                std::cout << message;
+            }
+            // Due to the problem faced in Linux environment which is described
+            // in https://gcc.gnu.org/ml/gcc/2003-12/msg00743.html, we could not use
+            // std::cout here. outstream flush() function in stdlib 3.4.4 does not handle pthread_cancel call
+            // appropriately.
+            std::flush(std::cout);
+        }
+
+        LeveledLogger::LeveledLogger(ILogger &logger, client::LoggerLevel::Level logLevel) : logger(logger) {
+            if (logger.isEnabled(logLevel)) {
+                logger.lockMutex.lock();
+                logger.printMessagePrefix(logLevel);
+            }
+        }
+
+        LeveledLogger::~LeveledLogger() {
+            std::cout << std::endl;
+            logger.lockMutex.unlock();
```

Comment:
The constructor takes the lock if only enabled.
The destructor always unlocks it.

Reply:
good finding 👍 Fixed it.